# Data Wrangling
## Various Group By Methods
### What we're going to learn
In this brief tutorial, we're going to learn three things. First, we're going to integrate pandas with a sql server. That way we can just pull data into our analysis and start working from there. Then we are going to learn how to perform a group by function in pandas. Finally we're going to see a second method of doing a group by. The second method leverages the fact that we are connected to a sql server.

### Preliminaries

In [7]:
import pandas
import mysql.connector
import mysettings

## Pandas Group By
### Connect to MySQL Server

In [12]:
cnx = mysql.connector.connect(user=mysettings.USER, password=mysettings.PASSWORD,
                              host=mysettings.HOST,
                              database=mysettings.DB)
sql='''
select
    *
from
    users
'''
df=pandas.io.sql.read_sql_query(sql,cnx)
df.head()

id  male  age
0   1     1   29
1   2     0   29
2   3     1   29
3   4     0   29
4   5     1   29

Group by the variable male using the pandas utility group by

In [16]:
df.groupby('male')['age'].mean()

male
0    33.4
1    32.6
Name: age, dtype: float64

At this point you see that the average age of women is 33.4, and that for men is 32.6. Okay that's fantastic. But why did we take a detour into the land of python before we performed our group by? We could have just pulled it into python already grouped. We are writing the sql query afterall.

## SQL Group By

In [14]:
sql='''
select
    male,
    avg(age) as age
from
    users
group by
    male
'''
df2=pandas.io.sql.read_sql_query(sql,cnx)
df2.head()

male   age
0     0  33.4
1     1  32.6

As you can see, you get the same answer as before but now our data comes in exactly the right form. So which method is better? I think it depends. If I needed to pull some data and do lots of analysis on it. I'd probably pull everything into a pandas dataframe and then use pandas group by function. It would allow me to get the results and preserve the data for additional analysis. But if I just needed to do a one-off qto find out the average age of men and women users analysis, I'd opt for the sql method. Although, this begs the question of why you would ever start out in python for that sort of analysis. You have to manually control the connection to the server, and stuff. In that situation, I would probably reach for a tool like phpmyadmin. Unless I couldn't and I was forced to use python, or if my dataset was huge and I likely would save time by calculating this group by on the sql server.

In [15]:
cnx.close()

PSA: Never leave your db connections open. You thought I forgot didn't you.

# If you liked this tutorial make sure to subscribe to my email list so that you can be the first to know when I post something new. You can do that at [BarnesAnalytics.com](http://www.barnesanalytics.com)